content/getting_started/CUDAGuide.md

```diff
@@ -0,0 +1,131 @@
+
+extern void __enzyme_autodiff(f_ptr,
+    int, double*, double*,
+    int, double*, double*);
+
+int main() {
+
+    double x = 1.4;
+    double d_x = 0.0;
+    double y;
+    double d_y = 1.0;
+
+    __enzyme_autodiff(foo,
+        enzyme_dup, &x, &d_x,
+        enzyme_dup, &y, &d_y);
+
+    printf("%f %f\n", x, y);
+    printf("%f %f\n", d_x, d_y);
+
+}
+```
+
+A one-liner compilation of the above using Enzyme:
+
+``` sh
+clang test2.cpp -Xclang -load -Xclang /path/to/ClangEnzyme-11.so -O2 -fno-vectorize -fno-unroll-loops
+```
+
+## CUDA Example
+
+When porting the above code, there are some caveats to be aware of:
+
+1. CUDA 10.1 is the latest supported CUDA at the time of writing (Jan/20/2021) for LLVM 11.
+2. ```__enzyme_autodiff``` should only be invoked on ```___device___``` code, not ```__global__``` kernel code. ```__global__``` kernels may be supported in the future.
+3. ```--cuda-gpu-arch=sm_xx``` is usually needed as the default ```sm_20``` is unsupported by modern CUDA versions.
+
+
+```cpp
+#include <stdio.h>
+
+void __device__ foo_impl(double* x_in, double *x_out) {
+    x_out[0] = x_in[0] * x_in[0];    
+}
+
+typedef void (*f_ptr)(double*, double*);
+
+extern void __device__ __enzyme_autodiff(f_ptr,
+    int, double*, double*,
+    int, double*, double*
+);
+
+void __global__ foo(double* x_in, double *x_out) {
+    foo_impl(x_in, x_out);
+}
+
+int __device__ enzyme_dup;
+int __device__ enzyme_out;
+int __device__ enzyme_const;
+
+void __global__ foo_grad(double* x, double *d_x, double *y, double *d_y) {
+
+    __enzyme_autodiff(foo_impl,
+        enzyme_dup, x, d_x,
+        enzyme_dup, y, d_y);
+
+}
+
+int main() {
+
+    double *x, *d_x, *y, *d_y; // device pointers
+
+    cudaMalloc(&x, sizeof(*x));
+    cudaMalloc(&d_x, sizeof(*d_x));
+    cudaMalloc(&y, sizeof(*y));
+    cudaMalloc(&d_y, sizeof(*d_y));
+
+    double host_x = 1.4;
+    double host_d_x = 0.0;
+    double host_y;
+    double host_d_y = 1.0;
+
+    cudaMemcpy(x, &host_x, sizeof(*x), cudaMemcpyHostToDevice);
+    cudaMemcpy(d_x, &host_d_x, sizeof(*d_x), cudaMemcpyHostToDevice);
+    cudaMemcpy(y, &host_y, sizeof(*y), cudaMemcpyHostToDevice);
+    cudaMemcpy(d_y, &host_d_y, sizeof(*d_y), cudaMemcpyHostToDevice);
+
+    // foo<<<1,1>>>(x, y); fwd-pass only
+    foo_grad<<<1,1>>>(x, d_x, y, d_y); // fwd and bkwd pass
+
+    cudaDeviceSynchronize(); // synchroniz
+
+    cudaMemcpy(&host_x, x, sizeof(*x), cudaMemcpyDeviceToHost);
+    cudaMemcpy(&host_d_x, d_x, sizeof(*d_x), cudaMemcpyDeviceToHost);
+    cudaMemcpy(&host_y, y, sizeof(*y), cudaMemcpyDeviceToHost);
+    cudaMemcpy(&host_d_y, d_y, sizeof(*d_y), cudaMemcpyDeviceToHost);
+
+    printf("%f %f\n", host_x, host_y);
+    printf("%f %f\n", host_d_x, host_d_y);
+
+}
+```
+
+The one-liner compilation step is (against sm_70):
+
+```sh
+clang test3.cu -Xclang -load -Xclang /path/to/ClangEnzyme-11.so -O2 -fno-vectorize -fno-unroll-loops -fPIC --cuda-gpu-arch=sm_70 -lcudart -L/usr/local/cuda-10.1/lib64
```

Comment:
Can you add a warning here that loading Enzyme into clang in this manner may not properly nest Enzyme between optimization passes, which may have a performance impact. Explicitly calling it via opt (in a similar manner to the regular usage guide) allows a proper nesting, but is more annoying to set up.

Reply:
I've actually tried to this explicitly, I noticed that two intermediate ll files get produced via

```
clang test3.cu -S -emit-llvm -O2 -fno-vectorize -fno-slp-vectorize -fno-unroll-loops --cuda-gpu-arch=sm_70 -fPIC
```
```
ls
test3.ll test3-cuda-nvptx64-nvidia-cuda-sm_70.ll
```
I followed the instructions in the Getting Started guide, and built them separately:

```
opt test3.ll -load=/home/yutong/Code/Enzyme/enzyme/build/Enzyme/LLVMEnzyme-11.so -enzyme -S -o test3_output.ll
opt test3_output.ll -load=/home/yutong/Code/Enzyme/enzyme/build/Enzyme/LLVMEnzyme-11.so -enzyme -o test3_opt.ll -S

opt test3-cuda-nvptx64-nvidia-cuda-sm_70.ll -load=/home/yutong/Code/Enzyme/enzyme/build/Enzyme/LLVMEnzyme-11.so -enzyme -S -o test3-cuda_output.ll
opt test3-cuda_output.ll -load=/home/yutong/Code/Enzyme/enzyme/build/Enzyme/LLVMEnzyme-11.so -enzyme -o test3-cuda_opt.ll -S
```

However, the final object creation step results in an error that I haven't seen before:

```
13:39 $ clang -c test3_opt.ll test3-cuda_opt.ll
warning: overriding the module target triple with x86_64-unknown-linux-gnu [-Woverride-module]
'sm_70' is not a recognized processor for this target (ignoring processor)
'+ptx64' is not a recognized feature for this target (ignoring feature)
'+sm_70' is not a recognized feature for this target (ignoring feature)
'sm_70' is not a recognized processor for this target (ignoring processor)
'sm_70' is not a recognized processor for this target (ignoring processor)
'+ptx64' is not a recognized feature for this target (ignoring feature)
'+sm_70' is not a recognized feature for this target (ignoring feature)
'sm_70' is not a recognized processor for this target (ignoring processor)
fatal error: error in backend: 64-bit code requested on a subtarget that doesn't support it!
clang-11: error: clang frontend command failed with exit code 70 (use -v to see invocation)
clang version 11.0.1 (https://github.com/llvm/llvm-project.git 43ff75f2c3feef64f9d73328230d34dac8832a91)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /home/yutong/local/bin
clang-11: note: diagnostic msg: Error generating preprocessed source(s) - no preprocessable inputs.
```

Reply:
Yeah this process is honestly really annoying to do manually (and in fact is why the ClangEnzyme.so was created). If you compile a regular cuda code with -v you'll see a bunch of steps necessary to successfully compile / separately link the cpu and cuda codes together. I don't think it's worth suggesting anyone take this path, but just giving a sentence explaining that there might be performance implications as a result (which can be remedied by either properly injective Enzyme using opt, or fine tuning how the ClangEnzyme loader works).

Reply:
Okay thanks - I'll do this in a later pass.